### PR TITLE
Fuel scooping rework

### PR DESCRIPTION
**Pull request by Mc-Pain**
Fuel scooping was:

1) based on random numbers
2) not aware of time acceleration
3) inefficient if rate was too high (which is still unlikely to happen without time acceleration)
4) not working properly (see https://github.com/pioneerspacesim/pioneer/issues/5237)

Instead of this, I suggest:
1) accumulate hydrogen at rate based on speed and pressure
2) replace strict speed/pressure requirement (I think it's likely being able to scoop even at sparse atmospheres even when flying fast enough)

### CODE REVIEW

**Comment by impaktor** [comment]
Thanks for looking into this!

Does this PR fix the issue you link to?

Commit (structure & messages) look nice. Not sure why you would prefer (slow) division over multiplication in the last commit, though?

**Comment by Mc-Pain** [comment]
> Does this PR fix the issue you link to?

Nope, since it's fixed already but there are other things I complained about

> Commit (structure & messages) look nice. Not sure why you would prefer (slow) division over multiplication in the last commit, though?

I'd prefer readability over obscurity.
`a / 300000` looks nicer than `a * 0.00000333`

How much impact slower (?) division would have in this case?
Are we sure if the compiler (gcc 13.2.1 in my case) won't optimize this?

It could replace dividing with multiplying by (1 / divisor) at compile time, and vice versa.

**Comment by impaktor** [comment]
> I will remove this commit unless it is okay to improve readability at cost of speed

he-he, @Web-eWorks gets final say. 

Also, it's not clear to me (from memory when I looked at the code a few hours ago), where that 300000 comes from, should it be a named constant? Doesn't matter much since it was that way from before. I'm just thinking loud.

**Comment by Mc-Pain** [comment]
> Also, it's not clear to me (from memory when I looked at the code a few hours ago), where that 300000 comes from

it comes from d66d2d14e621936529b7cc0208ca2e0eaaab5da6, where 0.00001f became 0.00000333f (scooping capacity for fuel scoop has been tripled so overall rate should remain unchanged)

**Comment by Mc-Pain** [comment]
@Web-eWorks what if I replace the last commit with one replacing 0.00000333f by constant?

```
// used for calculating a fuel scoop rate
const double hydrogen_pressure = 0.00000333f
```

**Comment by Web-eWorks** [comment]
Apologies for the delay in getting back around to this PR - you've probably seen what took up my time.

> what if I replace the last commit with one replacing 0.00000333f by constant?

As long as the constant has a comment indicating the unit conversion performed by employing the constant (i.e. `m/s * kg/m^3 -> kg of h2`) it's a good idea in my eyes. The main thing I'm concerned about here is that it's easy to follow the logical / physical process attempting to be modelled in this algorithm, so future contributors can understand and improve it if needed. Your current version is fairly easy to follow, but could perhaps be expressed in terms of minimum/total flow rate rather than using an intermediate speed*density product to determine whether scooping is happening.

Let me know when you've arrived on a model you're happy with and I'll review it for merge.

**Comment by Mc-Pain** [comment]
Now I'm happy enough, though I was not intended to rework the model completely.

Just the thing about scooping not happening when any of (speed, pressure) is below limits.
Perhaps it would be correct if we set a limit on their product instead and remove randomization of process (I am certain that H2 rate on such gas flow would not be fluctuating much)

Otherwise, imagine if such a place on Earth with sufficiently low oxygen rate appears randomly from natural causes (i.e. not including industrial pollution or closed space with a lot of humans without air vent)

**Comment by Web-eWorks** [comment]
> (I am certain that H2 rate on such gas flow would not be fluctuating much)

> Otherwise, imagine if such a place on Earth with sufficiently low oxygen rate appears randomly from natural causes (i.e. not including industrial pollution or closed space with a lot of humans without air vent)

If I remember correctly, on Earth a disproportionate quantity of O2 is concentrated in the troposphere and other elements behave the same in large gas giants, where there are striated layers of different types/mixtures of gases rather than a 100% homogeneous atmosphere.

That might be a useful thing to model in terms of variable H2 gas flow, if I'm reading your comment correctly. However, that would be a task for another follow-up PR well in the future when we have a more complex and complete atmospheric simulation with multiple atmospheric layers (and an implementation of scanning gameplay for users to become aware of atmospheric composition).

In the context of this PR, a uniform H2 flow rate controlled only by atmospheric density is fine by me.

Finally, something to look at for this or a future PR might be controlling the scoop rate by the `atmosOxidizing` SystemBody parameter, as that is intended to control whether the atmosphere is primarily comprised of hydrogen/ammonia/etc. or oxygen-based compounds. The parameter is not exactly scientific, but that's partially associated with the low number of users in the codebase.

**Comment by Mc-Pain** [comment]
Can I merge this or I should wait for review then?

File: src/Ship.cpp
Comment (Web-eWorks) [inline src/Ship.cpp:1271]: `Pi::game->GetTimeAccelRate()` is not the "full picture" when it comes to making this function update-rate independent - you should multiply by `timeStep` instead, as that is the fraction of simulated time elapsed during the update period. You may need to adjust the rate of fuel scooping, as the function is currently scooping ~60x as fast as it should.

**Review by Web-eWorks — CHANGES_REQUESTED** [review]
Generally, unless a change is a small single-line bugfix we prefer to at least take a cursory glance at the code and test the changed mechanic to ensure it works as intended. This means that if no community members do any testing and report back their findings, testing and review usually has to wait until a developer is finished with their in-flight projects and can "switch gears" to building and reviewing the branch.

I've not yet been able to test this branch, but I can review the code at the moment - I've noticed one issue, but otherwise the code looks quite clean to me.

File: src/Ship.cpp
Comment (Mc-Pain) [reply src/Ship.cpp]: > as the function is currently scooping ~60x as fast as it should.

I don't actually see why.
Before PR, scooping rate was compared against [0; 1) random number every tick, and if the rate is bigger, a ton of hydrogen is added to your cargo. I kept this rate unchanged.

Did a quick tests:
Scoop at 100 m/s on Saturn (I'll post save later) for 5 minutes.
Atmospheric pressure: ~51.8 atm

Before scooping: 29 tons
After scooping
Old algorithm, try 1: 257 tons (228t scooped)
Old algorithm, try 2: 238 tons (209t scooped)
Old algorithm, try 3: 259 tons (230t scooped)
New algorithm, without 4394d63d3: 239 tons (210t scooped)
New algorithm, without 4394d63d3, x10 acceleration: 239 tons (210t scooped)
New algorithm, with 4394d63d3: 33 tons (4t scooped)

With changes you requested scooping became ~60x times slower

Comment (Mc-Pain) [reply src/Ship.cpp]: Attaching the savefile

[scooping-test.gz](https://github.com/pioneerspacesim/pioneer/files/12642288/scooping-test.gz)

Comment (Web-eWorks) [reply src/Ship.cpp:1271]: I meant that you may need to increase the rate of scooping to (re-)create the intended flow of fuel. The function is currently scooping at a timestep-invariant rate, which will only function correctly so long as the game is updating at a fixed 60 ticks per second. If any change occurs to the frequency of timestep updates, the rate of fuel will change in a manner that is not consistent with the atmospheric density or speed.

The function should be modified to multiply by `timestep`, which you did in https://github.com/pioneerspacesim/pioneer/commit/4394d63d3a3554a0915d1cade9bc4e81a544668f, *and* the rate adjusted upwards by exactly 60x. I specifically called out that it was scooping 60x as fast as you would expect if you tried to do the math by hand using the existing constants in the code.

Comment (Mc-Pain) [reply src/Ship.cpp:1271]: > If any change occurs to the frequency of timestep updates

Is it intended to change?

Comment (Web-eWorks) [reply src/Ship.cpp:1271]: > Is it intended to change?

Assume yes. There's a proposal on IRC right now for adaptive substepping to improve ships crashing into the ground at high time acceleration. Whether or not we intend to change the timestep however, the correct way to express the algorithm is in terms of the timestep value.

**Review by Web-eWorks — APPROVED** [review]

**Comment by Web-eWorks** [comment]
Looks good to me, apologies for the merge delay.

**Pull request merged**
